shared/middleware/files_api.rb

```diff
@@ -355,4 +357,212 @@ def copy_file(endpoint, encrypted_channel_id, filename, source_filename)
+      end
+    end
+
+    # if we're also deleting a file (on rename), remove it from the manifest
+    manifest.reject! { |e| e['filename'].downcase == params['delete'].downcase } if params['delete']
+
+    # write the manifest (assuming the entry changed)
+    response = bucket.create_or_replace(encrypted_channel_id, 'manifest.json', manifest.to_json, params['files-version']) unless entry_is_unchanged
+
+    # delete a file if requested (same as src file in a rename operation)
+    bucket.delete(encrypted_channel_id, params['delete']) if params['delete']
+
+    # return the new entry info
+    new_entry_hash['filesVersionId'] = response.version_id
+    new_entry_hash.to_json
+  end
+
+  # POST /v3/files/<channel-id>/?version=<version-id>&project_version=<project-version-id>
+  #
+  # Create or replace a file. We use this method so that IE9 can still
+  # upload by posting to an iframe.
+  #
+  post %r{/v3/files/([^/]+)/$} do |encrypted_channel_id|
+    dont_cache
+    # though this is JSON data, we're making the POST request via iframe
+    # form submission. IE9 will try to download the response if we have
+    # content_type json
+    content_type 'text/plain'
+
+    bad_request unless request.POST['files'] && request.POST['files'][0]
+
+    file = request.POST['files'][0]
+
+    bad_request unless file[:filename] && file[:tempfile]
+
+    files_put_file(encrypted_channel_id, file[:filename], file[:tempfile].read)
+  end
+
+  #
+  # PUT /v3/files/<channel-id>/<filename>?version=<version-id>&project_version=<project-version-id>
+  #
+  # Create or replace a file. Optionally overwrite a specific version.
+  #
+  put %r{/v3/files/([^/]+)/([^/]+)$} do |encrypted_channel_id, filename|
+    dont_cache
+    content_type :json
+
+    if params['src'].nil?
+      # read the entire request before considering rejecting it, otherwise varnish
+      # may return a 503 instead of whatever status code we specify. Unfortunately
+      # this prevents us from rejecting large files based on the Content-Length
+      # header.
+      body = request.body.read
+    end
+
+    files_put_file(encrypted_channel_id, filename, body)
+  end
+
+  #
+  # DELETE /v3/files/<channel-id>/<filename>?files-version=<project-version-id>
+  #
+  # Delete a file. If filename is '*', all files are deleted.
+  #
+  delete %r{/v3/files/([^/]+)/([^/]+)$} do |encrypted_channel_id, filename|
+    dont_cache
+    content_type :json
+
+    owner_id, _ = storage_decrypt_channel_id(encrypted_channel_id)
+    not_authorized unless owner_id == storage_id('user')
+
+    # read the manifest
+    bucket = FileBucket.new
+    manifest_result = bucket.get(encrypted_channel_id, 'manifest.json')
+    if manifest_result[:status] == 'NOT_FOUND'
+      if filename == '*'
+        return { filesVersionId: "unused" }.to_json
+      else
+        not_found
+      end
+    end
+    manifest = JSON.load manifest_result[:body]
+
+    if filename == '*'
```

Comment:
We check `filename == '*'` two places in this method, and with the early returns it looks like the two branches have very little in common.  Should we extract this path into a helper, or even make it its own route `/v3/files/([^/]+)/*$`?